### PR TITLE
Ignore SEGV during profiler unwind on Unix

### PR DESCRIPTION
**Pull request by maleadt**
Unix equivalent of #4159. Unpolished, works for my use case (libcuda tripping up libunwind).

### CODE REVIEW

File: src/signals-unix.c [hunk 1]
```diff
@@ -667,9 +675,16 @@ static void *signal_listener(void *arg)
...
-                    bt_size_cur += rec_backtrace_ctx((uintptr_t*)bt_data_prof + bt_size_cur,
-                            bt_size_max - bt_size_cur - 1, signal_context);
+                    profiler_state = 1;
+                    getcontext(&profiler_uc);
+                    if (profiler_state == 1) {
+                        // Get backtrace data
+                        bt_size_cur += rec_backtrace_ctx((uintptr_t*)bt_data_prof + bt_size_cur,
+                                bt_size_max - bt_size_cur - 1, signal_context);
+                    } else {
+                        jl_safe_printf("WARNING: profiler attempt to access an invalid memory location\n");
```
[the remark this answers is not in the document]
Comment (maleadt) [reply src/signals-unix.c]: Just mimicking https://github.com/JuliaLang/julia/blob/614d91776d089a708861ce77f7d80a2914ffa20d/src/signals-mach.c#L468. I agree it's not an interesting message for users though, but I've only seen it print at most a couple of times during nontrivial (1-10s) profile invocations.

[the remark this answers is not in the document]
Comment (JeffBezanson) [reply src/signals-unix.c]: Ok.

**Comment by yuyichao** [comment]
We already have a mechanism for this. https://github.com/JuliaLang/julia/blob/614d91776d089a708861ce77f7d80a2914ffa20d/src/rtutils.c#L1023

**Comment by maleadt** [comment]
OK, changed the approach. Will squash on merge.

File: src/signals-unix.c [hunk 2]
```diff
@@ -225,6 +225,18 @@ static void segv_handler(int sig, siginfo_t *info, void *context)
     jl_ptls_t ptls = jl_get_ptls_states();
     assert(sig == SIGSEGV || sig == SIGBUS);
 
+    // if we're profiling, this segfault is likely caused by the unwinder.
+    // ignore the signal and jump back to where we came from.
+    if (running && ptls->safe_restore) {
```
Comment (yuyichao) [inline src/signals-unix.c]: Why do you need this? Does it not work with the condition below?

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: src/signals-unix.c (hunk 2 above)
Comment (maleadt) [reply src/signals-unix.c]: Just being conservative, only affecting the case where the profiler is running. Do it unconditionally then?

Comment (yuyichao) [reply src/signals-unix.c]: I mean you shouldn't need any code here in the segfault handler. Have you tested that it doesn't work without this but with the condition a few lines below?

Comment (yuyichao) [reply src/signals-unix.c]: (Or in another word, it is meant to be doing this unconditionally)

Comment (maleadt) [reply src/signals-unix.c]: Ah ok, no it doesn't work, looking closer it triggers a segfault in `jl_call_in_ctx` (via `jl_throw_in_ctx..., jl_stackovf_exception, ...)`).
Inferring from the function names, doesn't that behave differently from the plain `longjmp` I do here, would I need to catch an exception then somehow?

Comment (yuyichao) [reply src/signals-unix.c]: Hmm, segfault **in** `jl_call_in_ctx`? Did you get a `NULL` `context`?

Comment (yuyichao) [reply src/signals-unix.c]: Ah, I guess this thread doesn't have `signal_stack` allocated. I believe this should fix it (you can merge this with the falllback `ifdef` below if you want).

```patch
diff --git a/src/signals-unix.c b/src/signals-unix.c
index 0fafe121cd..8da89b5fc4 100644
--- a/src/signals-unix.c
+++ b/src/signals-unix.c
@@ -89,6 +89,14 @@ static void jl_call_in_ctx(jl_ptls_t ptls, void (*fptr)(void), int sig, void *_c
     // checks that the syscall is made in the signal handler and that
     // the ucontext address is valid. Hopefully the value of the ucontext
     // will not be part of the validation...
+    if (!ptls->signal_stack) {
+        sigset_t sset;
+        sigemptyset(&sset);
+        sigaddset(&sset, sig);
+        sigprocmask(SIG_UNBLOCK, &sset, NULL);
+        fptr();
+        return;
+    }
     uintptr_t rsp = (uintptr_t)ptls->signal_stack + sig_stack_size;
     assert(rsp % 16 == 0);
 #if defined(_OS_LINUX_) && defined(_CPU_X86_64_)
```

Comment (yuyichao) [reply src/signals-unix.c]: We don't handle segfault the same way on OSX. I don't really know if the two ways could be used together.

**Review by yuyichao — APPROVED** [review]

**Pull request merged**